File: sieve2.py
import math
print ( 'Python' )
Expected = 15485863
N, FINDPRIME = 16000000,1000000
count, N_sqrt = 1, int(math.sqrt(N))
b = []
for s in range(0,N):
  b.append(1)
for i in range(3, N_sqrt, 2):
  if b[i]:
    count = count + 1
    for k in range((i*i), N, i*2):
      b[k] = 0
for i in range(i, N, 2):
  if b[i]:
    count += 1
  if count >= FINDPRIME:
    break
print ( 'Total Primes = ', count )
print ( 'Max Primes = ', i )
if (i != Expected):
  print ( "ERROR: Largest Prime was incorrect. Should be: ", Expected )

File: sieve2.java
import java.util.BitSet;
import static java.lang.Math.sqrt;
import static java.lang.System.out;

public class sieve2 {

  public static void main(String[] s) {
    out.println("Java");
    int count = 1, i = 3;
    int N_sqrt = (int)sqrt((double)N);
    BitSet b = new BitSet(N);

    for (i = 3; i <= N_sqrt; i += 2) {
      if ( ! b.get(i) ) {
        count++;
        for (int k = i*i; k <= N; k += 2 * i) b.set(k);
      }
    }
    for (;i <= N; i += 2) {
      if ( ! b.get(i) ) {
        count++;
        if (count >= FINDPRIME) break;
      }
    }
    out.println("Total Primes = " + count);
    out.println("Max Prime = " + i);
    if (i != Expected) out.println("ERROR: Largest Prime was incorrect. Should be: " + Expected);
  }

  static final int N = 16000000;
  static final int FINDPRIME = 1000000;
  static final int Expected = 15485863;
}
